### PR TITLE
StreamField developer docs

### PR DESCRIPTION
**Pull request by gasman**

### CODE REVIEW

**Comment by davecranwell** [comment]
Looks good to me and certainly tallies with how I understand it all.

File: docs/core_components/pages/streamfield.rst [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        first_name = blocks.CharBlock(required=True)
+        surname = blocks.CharBlock(required=True)
+        photo = ImageChooserBlock()
+        biography = blocks.RichTextBlock()
+
+        class Meta:
+            template = 'myapp/blocks/person.html'
+            icon = 'user'
+
+
```
Comment (kaedroho) [inline docs/core_components/pages/streamfield.rst]: Spacing here is a little inconsistent

File: docs/core_components/pages/streamfield.rst [hunk 2]
```diff
@@ -0,0 +1,412 @@
+Freeform page content using StreamField
+=======================================
+
+Wagtail's content model encourages thinking about the information on a page in a formal, structured way; for example, a page about an event is not just an assortment of headings, paragraphs and images, but has specific fields for dates, location, a list of speakers and so on. This ensures that the template developer has a known, predictable set of fields to work with, and can present that information consistently throughout the site in the most appropriate arrangement. It also allows the information to be re-used and repurposed, such as displaying an "events near you" listing.
```
Comment (kaedroho) [inline docs/core_components/pages/streamfield.rst]: TBH, if I was reading this doc, I would skip these four paragraphs as it's quite a lot to read. But that's just me

Comment (davecranwell) [inline docs/core_components/pages/streamfield.rst]: I'm reluctantly inclined to second this, despite matthew's eloquent explanation of a difficult concept :P

Some of the background info could probably be removed and the rest abbreviated.

File: docs/core_components/pages/streamfield.rst [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    class BlogPage(Page):
+        author = models.CharField(max_length=255)
+        date = models.DateField("Post date")
+        body = StreamField([
+            ('heading', blocks.CharBlock(classname="full title")),
+            ('paragraph', blocks.RichTextBlock()),
+            ('image', ImageChooserBlock()),
+        ])
+
+    BlogPage.content_panels = [
```
Comment (kaedroho) [inline docs/core_components/pages/streamfield.rst]: Would it be neater to define content_panels in the class definition?

**Comment by kaedroho** [comment]
:+1: from me. Might be worth getting someone who's unfamilar with Streamfield to look at this.

File: docs/core_components/pages/streamfield.rst (hunk 2 above)
Comment (zerolab) [inline docs/core_components/pages/streamfield.rst]: I would rather have the long form, but with the suggested extra headings.

**Comment by zerolab** [comment]
The docs clarify a lot of the questions and uncertainties I had when I started implementing something with StreamField.

:+1:

File: docs/core_components/pages/streamfield.rst (hunk 2 above)
Comment (gasman) [inline docs/core_components/pages/streamfield.rst]: Happy to cut it down a bit. I wrote this part before I knew I'd be writing about streamfield for the torchbox.com blog, so I was probably leaping on my soapbox a bit too enthusiastically :-) Would probably make sense to have a short and to-the-point introduction here, and link to the blog post for further reading.

**Comment by gasman** [comment]
Intro section updated - a bit less rambly now :-)

**Comment by gasman** [comment]
Rebased on top of #1090.

**Pull request merged**